app/js/lib/foxbox/webpush.js

```diff
@@ -0,0 +1,104 @@
+'use strict';
+
+// Private members
+const p = Object.freeze({
+  // Properties,
+  net: Symbol('net'),
+  settings: Symbol('settings'),
+
+  // Methods:
+  listenForMessages: Symbol('listenForMessages'),
+  dispatchEvent: Symbol('dispatchEvent')
+});
+
+export default class WebPush {
+  constructor(network, settings, callback) {
+    this[p.net] = network;
+    this[p.settings] = settings;
+    this.callback = callback;
+
+    Object.seal(this);
+  }
+
+  subscribeToNotifications(resubscribe = false) {
+    const settings = this[p.settings];
+    const boxPath = 
+      `${this[p.net].origin}/api/v${this[p.settings].apiVersion}`;  
+
+    if (!navigator.serviceWorker) {
+      return Promise.reject('No service worker supported');
+    }
+
+    navigator.serviceWorker.addEventListener('message',
+      this[p.listenForMessages].bind(this));
+
+    if (settings.pushEndpoint && settings.pushPubKey && !resubscribe) {
+      return Promise.resolve();
+    }
+
+    return navigator.serviceWorker.ready
+      .then((reg) => {
+        reg.pushManager.subscribe({userVisibleOnly: true})
+      .then((subscription) => {
+        const endpoint = subscription.endpoint;
+        const key = subscription.getKey ? subscription.getKey('p256dh') : '';
+        settings.pushEndpoint = endpoint;
+        settings.pushPubKey = btoa(String.fromCharCode.apply(null,
+          new Uint8Array(key)));
+
+        // Send push information to the server
+        // @todo: We will need some library to write taxonomy messages
+        const pushConfigurationMsg = [[
+            [{
+              id: 'setter:subscribe.webpush@link.mozilla.org'
+            }], {
+              Json: {
+                subscriptions: [{
+                  public_key: settings.pushPubKey,
+                  push_uri: settings.pushEndpoint
+                }]
+              }
+            }
+          ]];
+
+        return this[p.net].fetchJSON(boxPath + '/channels/set',
+          'PUT', pushConfigurationMsg)
+        .then(() => {
+          // Setup some common push resources
+          const pushResourcesMsg = [[
+              [{
+                id: 'setter:resource.webpush@link.mozilla.org'
+              }], {
+                Json: {
+                  resources: ['res1']
+                }
+              }
+
+            ]];
+          return this[p.net].fetchJSON(boxPath + '/channels/set',
+           'PUT', pushResourcesMsg);
+        });
+      })
+      .catch((error) => {
+        if (Notification.permission === 'denied') {
+          return Promise.reject('Permission request was denied.');
+        } else {
+          console.error('Error while saving subscription ', error);
+          return Promise.reject('Subscription error: ', error);
```

Comment:
In this case I'm catching just to collect some information on why the push registration won't worked